### PR TITLE
Systematic HPO: Refactor

### PR DESCRIPTION
**Pull request by TammoR**
*Description of changes:*
Refactoring of this pull request: https://github.com/awslabs/datawig/pull/62
Thanks for the initial feedback. I've now moved HPO into it's own module.

I close the old pull request because the new changes will be easier to track here.
The inner workings stay mostly unchanged. I copy below the comments on the current limitations of this implementation:

_Let me be specific on what this was set out to do. I start with illustrating why this is difficult.
We want to be able to optimise all configurations of hyperparameters. This includes parameters with conditional dependencies. E.g. one hyperparameter for a particular column could be the column type: (categorical, string). If the column type is string, a next hyperparameter (that applies only conditionally on the previous one) is the embedding type (Tfidf, Hashing, LSTM). Each of these can again have a different set of conditional hyperparameters, and so on. There would be yet another parameter on top, that is the option to have multiple embeddings for the same column._

_It's natural to have a multi-level nested data-structure to store the different parameter values or distributions to draw from. This is done in the proposed implementation, but to keep things simple there are some limitations listed in the following._

_Only one string column embedding strategy is supported. This is hard coded (hashing vectoriser at the moment. Maybe swap for tfidf)
If the same column is embedded multiple times, these embeddings share all hyperparameters except for the token type (chars, words).
All input columns have a separate column encoder. There is no option to combine different columns into one (other than through pre-processing, which is easy enough)
There is no mechanism to automatically distinguish text from categorical columns. For categorical encodings, the user needs to supply the column type as parameter.
The same column can not be encoded as multiple types. E.g. string and categorical.
No support for drawing from distributions. Values need to be supplied as list.
I'd be interested in whether you think this is a reasonable compromise._

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: datawig/mxnet_input_symbols.py [hunk 1]
```diff
@@ -102,15 +103,15 @@ class LSTMFeaturizer(Featurizer):
     def __init__(self,
                  field_name: str,
                  seq_len: int = 500,
-                 vocab_size: int = 40,
```
Comment (felixbiessmann) [inline datawig/mxnet_input_symbols.py]: I'd rather not introduce backwards incompatible breaking changes, if not absolutely required. These names were changed a couple times back and forth and it's really a matter of taste what people think is most intuitive. In fact one of the initial choices for this parameter has been ``max_tokens``

File: datawig/mxnet_input_symbols.py [hunk 2]
```diff
@@ -151,11 +152,11 @@ class EmbeddingFeaturizer(Featurizer):
 
     def __init__(self,
                  field_name: str,
-                 vocab_size: int = 100,
```
Comment (felixbiessmann) [inline datawig/mxnet_input_symbols.py]: Same here

File: datawig/simple_imputer.py [hunk 3]
```diff
@@ -184,7 +183,7 @@ def fit_hpo(self,
         """
 
         if weight_decay is None:
-            weight_decay = [0]
+            weight_decay = [0, 1e-4, 1e-3, 1e-2]
```
Comment (felixbiessmann) [inline datawig/simple_imputer.py]: are these parameter candidates sensible? i mean have you seen improvements in the majority of experiments with any of those? if now, they'll merely multiply your training time by 4 which can be significant.

File: datawig/mxnet_input_symbols.py (hunk 1 above)
Comment (TammoR) [reply datawig/mxnet_input_symbols.py]: Backwards compatibility is a good point. It's not a matter of taste, however, but of consistency.  max_tokens and vocab_size are the same thing. BowEncoder takes max_tokens as input, BowFeaturizer takes vocab_size. In fact these two parameter should always be identical, right?

File: datawig/simple_imputer.py (hunk 3 above)
Comment (TammoR) [reply datawig/simple_imputer.py]: The fit_hpo method should actually be removed. It's replaced by the new module. So no need to discuss this.

Comment (felixbiessmann) [reply datawig/simple_imputer.py]: hm. not sure whether fit_hpo should be removed. but my point was about the parameter choices. have you seen any improvement with any of those values in experiments?

File: datawig/mxnet_input_symbols.py (hunk 1 above)
Comment (felixbiessmann) [reply datawig/mxnet_input_symbols.py]: I agree it's not consistent - but all these names were cr'd previously and I remember that correctly this inconsistency was introduced to make things more consistent amongst the Featurizer classes. 

I'm just trying to say that if we break the API we should really have good reasons for this.

**Comment by TammoR** [comment]
What I tried to convey in my introductory comment is exactly aligned with your pros/cons on comprehensively exposing HPs. The solution I propose is intended to hit a sweet spot in the middle between the two extremes.

I'm afraid there is a basic misunderstanding as to what this PR does. Sorry if it was not written clearly enough.
The proposed changes do not _add any computational complexity_ nor do they _decrease usability_. 
They add additional functionality to learn HPs. The usability of all previously available features remains unchanged. This PR does not by any means, _remove the usability of the `SimpleImputer`_.

It was intended to be a contribution to the _discussions and refactorings by the entire team_.

The main purpose of this pull request was to gather feedback on the features and the limitations of the proposed implementation.
I have not put extensive thought into how to best design the outside-facing API, since this is best done once I've gathered sufficient feedback on the internal workings. (It could remain part of the `SimpleImputer` class, as in the earlier pull request, or be instantiated through a call to some method of `SimpleImputer`, etc.)

File: datawig/mxnet_input_symbols.py (hunk 2 above)
Comment (TammoR) [reply datawig/mxnet_input_symbols.py]: resolving for brevity. this occurs many times throughout.

File: datawig/mxnet_input_symbols.py (hunk 1 above)
Comment (TammoR) [reply datawig/mxnet_input_symbols.py]: Alternative I would write a workaround and internally assign the parameter names to comply with the inconsistent naming scheme. Happy to do that. 

I can not see any reason for _consistency among feature classes_ (?)

File: datawig/column_encoders.py [hunk 4]
```diff
@@ -440,19 +440,24 @@ def __init__(self,
                  output_column: str = None,
                  max_tokens: int = 2 ** 18,
                  tokens: str = 'chars',
+                 ngram_range: tuple = None,
```
Comment (felixbiessmann) [inline datawig/column_encoders.py]: nice

File: datawig/column_encoders.py [hunk 5]
```diff
@@ -440,19 +440,24 @@ def __init__(self,
...
         elif tokens == 'chars':
-            self.vectorizer = TfidfVectorizer(max_features=int(self.output_dim), ngram_range=(1, 5),
-                                                analyzer="char")
+            self.vectorizer = TfidfVectorizer(max_features=int(self.output_dim), ngram_range=ngram_range,
+                                              analyzer="char")
         else:
             logger.info(
                 "BowEncoder attribute tokens has to be 'words' or 'chars', defaulting to 'chars'")
-            self.vectorizer = TfidfVectorizer(max_features=int(self.output_dim), ngram_range=(1, 5), analyzer="char")
+            self.vectorizer = TfidfVectorizer(max_features=int(self.output_dim),
```
Comment (felixbiessmann) [inline datawig/column_encoders.py]: not sure, maybe the int casting as part of the constructor?

File: datawig/hpo.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+            # TODO: add support for user provided functions
+        for uds in user_defined_scores:
+            hp[uds[1]] = uds[0](true=true, predicted=predicted, confidence=confidence)
+
+        return hp
+
+    def tune(self,
+             train_df: pd.DataFrame,
+             test_df: pd.DataFrame = None,
+             strategy: str = 'random',
```
Comment (felixbiessmann) [inline datawig/hpo.py]: great to have that feature!

File: datawig/hpo.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        Do grid search or random search for hyperparameter configurations. This method can not tune tfidf vs hashing
+        vectorization but uses tfidf. Also parameters of the output column encoder are not tuned.
+        
+        :param train_df: training data as dataframe
+        :param test_df: test data as dataframe; if not provided, a ratio of test_split of the
+                          training data are used as test data
+
+        :param strategy: 'random' for random search or 'grid' for exhaustive search
+        :param num_evals: number of evaluations for random search
+        :param user_defined_scores: list with entries (Callable, str), where callable is a function
```
Comment (felixbiessmann) [inline datawig/hpo.py]: maybe add one example, or list the default scoring?

File: datawig/column_encoders.py (hunk 5 above)
Comment (TammoR) [reply datawig/column_encoders.py]: yeah, good call.

File: datawig/column_encoders.py (hunk 4 above)
Comment (tdhd) [inline datawig/column_encoders.py]: please add `ngram_range` to the docstring above

File: datawig/column_encoders.py [hunk 8]
```diff
@@ -539,19 +544,23 @@ def __init__(self,
                  output_column: str = None,
                  max_tokens: int = 2 ** 18,
                  tokens: str = 'chars',
+                 ngram_range: tuple = None,
```
Comment (tdhd) [inline datawig/column_encoders.py]: please add `ngram_range` to the docstring above

File: datawig/evaluation.py [hunk 9]
```diff
@@ -80,7 +80,7 @@ def evaluate_and_persist_metrics(true_labels_string,
 
         for att in true_labels_string.keys():
             metrics[att]['precision_recall_curves'] = {}
-            for label in range(1, predictions_proba[att].shape[-1] + 1):
+            for label in range(1, predictions_proba[att].shape[-1]):
```
Comment (tdhd) [inline datawig/evaluation.py]: this is already addressed in https://github.com/awslabs/datawig/pull/67 ?

File: datawig/hpo.py [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+
+from pandas.api.types import is_numeric_dtype
+from sklearn.metrics import mean_squared_error, f1_score, precision_score, accuracy_score, recall_score
+
+from .column_encoders import BowEncoder, CategoricalEncoder, NumericalEncoder, ColumnEncoder, TfIdfEncoder
+from .mxnet_input_symbols import BowFeaturizer, NumericalFeaturizer, Featurizer, EmbeddingFeaturizer
+from .utils import logger, get_context, random_split, rand_string, flatten_dict, merge_dicts
+
+
+class HPO:
```
Comment (tdhd) [inline datawig/hpo.py]: `class _HPO:` to indicate this is for internal use only

File: datawig/hpo.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+    """
+    Implements systematic hyperparameter optimisation for datawig
+
+    Example usage:
+
+    imputer = SimpleImputer(input_columns, output_column)
+    hps = dict( ... )  # specify hyperparameter choices
+    hpo = HPO(impter, hps)
```
Comment (tdhd) [inline datawig/hpo.py]: `_HPO`

File: datawig/simple_imputer.py [hunk 12]
```diff
@@ -516,7 +432,37 @@ def load(output_path: str) -> Any:
...
+        """
+        Load model after hyperparameter optimisation has ran. Overwrite local artifacts of self.imputer.
+
+        :param hpo_idx: Index of the model to be loaded. Default,
+                    load model with highest weighted precision or mean squared error.
+
+        :return: imputer object
+        """
+
+        if hpo_idx is None:
```
Comment (tdhd) [inline datawig/simple_imputer.py]: what if hpo has not run before?

File: datawig/simple_imputer.py [hunk 13]
```diff
@@ -516,7 +432,37 @@ def load(output_path: str) -> Any:
...
+            else:
+                hpo_idx = self.hpo.results['precision_weighted'].astype(float).idxmax()
+            logger.info("Selecting imputer with maximum weighted precision.")
+
+        # copy artifacts from hp run to self.output_path
+        imputer_dir = self.output_path + str(hpo_idx)
+        files_to_copy = glob.glob(imputer_dir + '/*.json') +\
+                        glob.glob(imputer_dir + '/*.pickle') +\
+                        glob.glob(imputer_dir + '/*.params')
+        for file in files_to_copy:
```
Comment (tdhd) [inline datawig/simple_imputer.py]: `file` is a builtin, rename to something like `file_name`?

File: datawig/hpo.py [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+        if is_numeric_dtype(train_df[simple_imputer.output_column]):
+            label_column = [NumericalEncoder(simple_imputer.output_column)]
+            logger.info("Assuming numeric output column: {}".format(simple_imputer.output_column))
+        else:
+            label_column = [CategoricalEncoder(simple_imputer.output_column)]
+            logger.info("Assuming categorical output column: {}".format(simple_imputer.output_column))
+
+        global_parms = {key.split(':')[1]: val for key, val in hp.iteritems() if 'global' in key}
+
+        from . import Imputer  # needs to be imported here to avoid circular dependency
```
Comment (tdhd) [inline datawig/hpo.py]: can you move that to the beginning of `__fit_hp` please?

File: datawig/hpo.py [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+
+        for hp_idx, (_, hp) in enumerate(hps_flat.iterrows()):
+            if elapsed_time > max_running_hours:
+                logger.info('Finishing hpo because max running time was reached.')
+                break
+
+            # if concat_columns True, set all n.a. hps to n.a. TODO
+            logger.info("Fitting hpo iteration " + str(hp_idx) + " with parameters\n\t" +
+                        '\n\t'.join([str(i) + ': ' + str(j) for i, j in hp.items()]))
+            name = hpo_run_name + str(hp_idx)
```
Comment (tdhd) [inline datawig/hpo.py]: you are reusing how the name is constructed in `simple_imputer.load_hpo_model`. maybe it is better to create the name in a function here and call it function also from `simple_imputer.load_hpo_model`.

File: datawig/simple_imputer.py [hunk 16]
```diff
@@ -516,7 +432,37 @@ def load(output_path: str) -> Any:
...
+
+        if hpo_idx is None:
+            if self.output_type == 'numeric':
+                hpo_idx = self.hpo.results['mse'].astype(float).idxmax()
+            else:
+                hpo_idx = self.hpo.results['precision_weighted'].astype(float).idxmax()
+            logger.info("Selecting imputer with maximum weighted precision.")
+
+        # copy artifacts from hp run to self.output_path
+        imputer_dir = self.output_path + str(hpo_idx)
```
Comment (tdhd) [inline datawig/simple_imputer.py]: call function in `hpo` to get the name of the given index

File: test/test_simple_imputer.py [hunk 17]
```diff
@@ -254,29 +261,181 @@ def test_imputer_hpo_text(test_dir, data_frame):
...
+        output_path=output_path
+    )
+
+    # Define default hyperparameter choices for each column type (string, categorical, numeric)
+
+    # hps = {}
+    # hps['global'] = {}
+    # hps['string_feature'] = {}
+    # hps['categorical_feature'] = {}
+    # hps['numeric_feature'] = {}
```
Comment (tdhd) [inline test/test_simple_imputer.py]: can this go?

File: datawig/hpo.py [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+
+        return hp_df_from_dict(flatten_dict(self.hps))
+
+    def __fit_hp(self,
+                 train_df: pd.DataFrame,
+                 test_df: pd.DataFrame,
+                 hp: pd.Series,
+                 simple_imputer,
+                 user_defined_scores: list = None,
+                 name: str = ''):
```
Comment (tdhd) [inline datawig/hpo.py]: if name is an empty string we're overwriting the results of each of the hpo runs? maybe it's better to remove the default value and check that it is not empty.

File: datawig/hpo.py (hunk 6 above)
Comment (tdhd) [reply datawig/hpo.py]: :+1:

File: datawig/hpo.py (hunk 7 above)
Comment (tdhd) [reply datawig/hpo.py]: is it possible to change the user defined function signatures to take parameters named, as in `udf(true, predicted, confidence)`?

File: datawig/simple_imputer.py [hunk 19]
```diff
@@ -133,225 +139,135 @@ def check_data_types(self, data_frame: pd.DataFrame) -> None:
...
+        :param ctx: List of mxnet contexts (if no gpu's available, defaults to [mx.cpu()])
+            User can also pass in a list gpus to be used, ex. [mx.gpu(0), mx.gpu(2), mx.gpu(4)]
+            This parameter is deprecated.s
 
-        self.hpo_results = hpo_results
+        :return: pd.DataFrame with with hyper-parameter configurations and results
+        """
 
+        if hpo_run_name is None:
+            hpo_run_name = ""
```
Comment (tdhd) [inline datawig/simple_imputer.py]: you could pass this also as `None` to `self.hpo.tune` and check it there

File: datawig/hpo.py [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+
+    def tune(self,
+             train_df: pd.DataFrame,
+             test_df: pd.DataFrame = None,
+             hps: dict = None,
+             strategy: str = 'random',
+             num_evals: int = 10,
+             max_running_hours: float = 96,
+             user_defined_scores: list = None,
+             hpo_run_name: str = '',
```
Comment (tdhd) [inline datawig/hpo.py]: default here could be `None` to indicate it is optional instead of an empty string.

File: test/test_simple_imputer.py [hunk 21]
```diff
@@ -25,6 +25,7 @@
 import pytest
 from sklearn.metrics import f1_score, mean_squared_error
 
+# from datawig.hpo import HPO
```
Comment (tdhd) [inline test/test_simple_imputer.py]: can this go?

File: datawig/hpo.py (hunk 15 above)
Comment (TammoR) [reply datawig/hpo.py]: In `simple_imputer.load_hpo_model`, the name is read directly from the hpo_results table and not constructed in the same way as in `hpo.tune()`, where the user defined name is combined with an index. I agree though, that the construction of the name is not very elegant.

**Comment by TammoR** [comment]
Regarding ea418a9
When drawing random hpo configurations for many different columns the parameter space can easily explode and it's not possible to materialize all hpo configurations. Using an iterator (e.g. from python itertools) does not allow us to draw random samples from the hpo configurations.
I have therefore implemented the function `utils.random_cartesian_product`. It does exactly that: Given some list of iterables it returns a given number of random samples from their full outer cartesian product without the need to materialize the entire product.

File: datawig/utils.py [hunk 22]
```diff
@@ -302,4 +318,53 @@ def generate_df_numeric(num_samples: int = 100,
...
-    })
\ No newline at end of file
+    })
+
+
+def random_cartesian_product(sets, num=10):
+    """
+    Return num random samples from the cartesian product of all iterators in sets.
+    Returns at most as many results as unique products exist.
+    """
```
Comment (tdhd) [inline datawig/utils.py]: Please make this a standard docstring, with `:param ...`. An example input/output of this function would help understand it better.

File: datawig/utils.py [hunk 23]
```diff
@@ -302,4 +318,53 @@ def generate_df_numeric(num_samples: int = 100,
...
+    for idx in idxs:
+        out.append(sample_cartesian(sets, idx, N))
+
+    return out
+
+
+def sample_cartesian(sets, idx, N):
+    """
+    Draw a single sample from the cartesian product with index idx
+    """
```
Comment (tdhd) [inline datawig/utils.py]: same comment as for `random_cartesian_product`

File: datawig/utils.py [hunk 24]
```diff
@@ -45,6 +46,20 @@
 logger.setLevel("INFO")
 
 
+def flatten_dict(d, parent_key='', sep=':'):
+    """
+    Flatten a nested dictionary and create new keys by concatenation
+    """
```
Comment (tdhd) [inline datawig/utils.py]: same comment as for `random_cartesian_product`

File: datawig/simple_imputer.py [hunk 25]
```diff
@@ -592,7 +506,41 @@ def load(output_path: str) -> Any:
...
+        imputer_dir = self.output_path + str(hpo_name)
+        files_to_copy = glob.glob(imputer_dir + '/*.json') + \
+                        glob.glob(imputer_dir + '/*.pickle') + \
+                        glob.glob(imputer_dir + '/*.params')
+        for file_name in files_to_copy:
+            shutil.copy(file_name,  self.output_path)
+
+        self.imputer = Imputer.load(self.output_path)
+
+        return
```
Comment (tdhd) [inline datawig/simple_imputer.py]:
```suggestion
```

File: datawig/_hpo.py [hunk 26]
```diff
@@ -0,0 +1,431 @@
...
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
+
+DataWig HPO
+Implements hyperparameter optimisation for datawig
+
+
```
Comment (tdhd) [inline datawig/_hpo.py]:
```suggestion
```

File: datawig/_hpo.py [hunk 27]
```diff
@@ -0,0 +1,431 @@
...
+        if test_df is None:
+            train_df, test_df = random_split(train_df, [.8, .2])
+
+        # process_hp_configurations(hps) and return random configurations
+        hps_flat = self.__preprocess_hps(train_df, simple_imputer, num_evals)
+
+        # # sample configurations for random search
+        # if strategy == 'random':
+        #     hps_flat = pd.concat([hps_flat_iter() for i in range(min(num_evals, max_configs))], axis=1).transpose()
+        #     # hps_flat = hps_flat.sample(n=min([num_evals, hps_flat.shape[0]]), random_state=10)
```
Comment (tdhd) [inline datawig/_hpo.py]: can this go?

**Pull request merged**